crates/iceberg/src/spec/transform.rs

```diff
@@ -261,6 +269,323 @@ impl Transform {
+                PrimitiveLiteral::Long(v) => Some(Datum::long(v + 1)),
+                PrimitiveLiteral::Decimal(v) => Some(Datum::decimal(v + 1)?),
+                PrimitiveLiteral::Date(v) => Some(Datum::date(v + 1)),
+                PrimitiveLiteral::Timestamp(v) => Some(Datum::timestamp_micros(v + 1)),
+                _ => Some(datum.to_owned()),
+            },
+            PredicateOperator::Eq
+            | PredicateOperator::LessThanOrEq
+            | PredicateOperator::GreaterThanOrEq
+            | PredicateOperator::StartsWith
+            | PredicateOperator::NotStartsWith => Some(datum.to_owned()),
+            _ => None,
+        };
+
+        Ok(adjusted_boundary)
+    }
+
+    /// Adjusts the comparison operator based on the specified datum and an
+    /// optional width constraint.
+    ///
+    /// This function modifies the comparison operator for `LessThan` and
+    /// `GreaterThan` cases to their inclusive counterparts (`LessThanOrEq`,
+    /// `GreaterThanOrEq`) unconditionally. For `StartsWith` and
+    /// `NotStartsWith` operators acting on string literals, the operator may
+    /// be adjusted to `Eq` or `NotEq` if the string length matches the
+    /// specified width, indicating a precise match rather than a prefix
+    /// condition.
+    fn adjust_operator(
+        op: &PredicateOperator,
+        datum: &Datum,
+        width: Option<u32>,
+    ) -> Option<PredicateOperator> {
+        match op {
+            PredicateOperator::LessThan => Some(PredicateOperator::LessThanOrEq),
+            PredicateOperator::GreaterThan => Some(PredicateOperator::GreaterThanOrEq),
+            PredicateOperator::StartsWith => match datum.literal() {
+                PrimitiveLiteral::String(s) => {
+                    if let Some(w) = width {
+                        if s.len() == w as usize {
+                            return Some(PredicateOperator::Eq);
+                        };
+                    };
+                    Some(*op)
+                }
+                _ => Some(*op),
+            },
+            PredicateOperator::NotStartsWith => match datum.literal() {
+                PrimitiveLiteral::String(s) => {
+                    if let Some(w) = width {
+                        let w = w as usize;
+
+                        if s.len() == w {
+                            return Some(PredicateOperator::NotEq);
+                        }
+
+                        if s.len() < w {
+                            return Some(*op);
+                        }
+
+                        return None;
+                    };
+                    Some(*op)
+                }
+                _ => Some(*op),
+            },
+            _ => Some(*op),
+        }
+    }
+
+    /// Adjust projection for temporal transforms, align with Java
+    /// implementation: https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java#L275
+    fn adjust_time_projection(
```

Comment:
Seems this function is used to fix the bug from earlier iceberg. But I don't see same thing in pyiceberg transform. Do we really need this fix?  cc @liurenjie1024 @Fokko

Reply:
> Is it breaking new things?

I have doubts because this does not exist in iceberg-python but it's a fixed logic. In java, also there is this process. I'm not sure whether we don't need it or iceberg-python miss it.